src/main/java/school/hei/haapi/repository/dao/ExamDao.java

```diff
@@ -0,0 +1,85 @@
+package school.hei.haapi.repository.dao;
+
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.criteria.*;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import lombok.AllArgsConstructor;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.query.QueryUtils;
+import org.springframework.stereotype.Repository;
+import school.hei.haapi.model.AwardedCourse;
+import school.hei.haapi.model.Course;
+import school.hei.haapi.model.Exam;
+import school.hei.haapi.model.Group;
+
+@Repository
+@AllArgsConstructor
+public class ExamDao {
+  private final EntityManager entityManager;
+
+  public List<Exam> findByCriteria(
+      Pageable pageable,
+      String title,
+      String courseCode,
+      String groupRef,
+      Instant examinationDateStart,
+      Instant examinationDateEnd,
+      String awardedCourseId) {
+    CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+    CriteriaQuery<Exam> query = builder.createQuery(Exam.class);
+    Root<Exam> root = query.from(Exam.class);
+    ArrayList<Predicate> predicates = new ArrayList<>();
+
+    Join<Exam, AwardedCourse> awardedCourseJoin = root.join("awardedCourse", JoinType.LEFT);
+    Join<AwardedCourse, Course> courseJoin = awardedCourseJoin.join("course", JoinType.LEFT);
+    Join<AwardedCourse, Group> groupJoin = awardedCourseJoin.join("group", JoinType.LEFT);
```

Comment:
Only instantiate the join class when needed, otherwise it will cause paginations issues